File: p2.py
#Time: O(mn)
#Space: O(mn)
class Solution:
    def isMatch(self, s, p):
        dp = [[False for _ in range(len(p)+1)] for _ in range(len(s)+1)]
        dp[0][0] = True

        for i in range(1,len(dp[0])):
            if i >1 and p[i-1] == '*':
                dp[0][i] = dp[0][i-2]

        for i in range(1, len(dp)):
            for j in range(1, len(dp[0])):
                if p[j-1] != '*':
                    if s[i-1] == p[j-1] or p[j-1] == '.':
                        dp[i][j] = dp[i-1][j-1]
                else:
                     
                    dp[i][j] = dp[i][j-2]

                    if p[j-2] == '.' or s[i-1] == p[j-2]:
                        if dp[i-1][j]:
                            dp[i][j] = dp[i-1][j] 

        return dp[-1][-1]

File: p2.java
// Time Complexity : O(m*n)
// Space Complexity :O(m*n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach

class Solution {
    public boolean isMatch(String s, String p) {
        boolean[][] dp = new boolean[s.length()+1][p.length()+1];
        dp[0][0] = true;
        //Fill the first row where we just have to check for the * case that too zero case
        for(int j=1; j<p.length()+1; j++){
            if(p.charAt(j-1) == '*'){
                dp[0][j] = dp[0][j-2];
            }
        }

        for(int i=1; i<s.length()+1; i++){
            for(int j=1; j<p.length()+1; j++){
                //If incoming characters match, get the daigonal up
                if(p.charAt(j-1) == s.charAt(i-1) || p.charAt(j-1) == '.'){
                    dp[i][j] = dp[i-1][j-1];
                }
                //If we have a *
                if(p.charAt(j-1) == '*'){
                    //If the previous element match, then we have one case and zero case
                    if(p.charAt(j-2) == s.charAt(i-1) || p.charAt(j-2) == '.'){
                        dp[i][j] = dp[i-1][j] || dp[i][j-2] ;
                    }
                    //If previous element do not match, we have only zero case
                    else{
                        dp[i][j] = dp[i][j-2] ;
                    }
                }
            }
        }

        return dp[s.length()][p.length()];
    }
}